alphagov/paas-bootstrap
# [#152065618] Upgrade Concourse to 3.8.0

## What

Upgrade Concourse to 3.8.0.

This is the last step of the Concourse upgrade process from 3.4.1 -> 3.8.0.

Following the upgrade notes written in https://concourse.ci/downloads.html#v370.

This PR contains the following changes:
 - extend the secret generator to generate BOSH 2.0 compatible variables, like SSH and RSA (see https://bosh.io/docs/variable-types.html). I started to use the net/ssh library for this (it's already a dependency of bosh-cli), so we have to run bundle install before we generate the secrets.
 - generating new SSH and RSA secrets for Concourse components
 - added new mandatory properties to atc, tsa and groundcrew
 - Upgrade the bootstrap Docker image versions to 3.8.0

Notes:
 - The https://bosh.io/docs/variable-types.html is not correct as the public SSH key is not in PEM format but in "ssh-rsa ..." format
 - The 3.7.0 upgrade instructions missing the tsa.public_host_key property which I found in https://github.com/concourse/concourse-deployment/blob/master/lite/concourse.yml#L91.

## How to review

❗️ This PR also contains https://github.com/alphagov/paas-bootstrap/pull/110 which has to be merged first.

1. Update your pipeline from this branch
    ```BRANCH=152065618_upgrade_concourse_4 make dev pipelines```
2. Run the create-bosh-concourse pipeline (after it reaches the concourse-deploy task you have to reload Concourse after a couple of minutes)
3. Check if Concourse works by uploading the pipelines again and running create-bosh-concourse again

I also ran the create-cloudfoundry pipeline as a test successfully, I leave it to you if you want to test it or not.

To test the bootstrap Concourse:
1. Bootstrap your deployer Concourse and run the create-bosh-concourse pipeline from there. You have to set the branch:
   ```BRANCH=152065618_upgrade_concourse_4 make dev deployer-concourse bootstrap```

I also tested the build Concourse bootstrap successfully, I leave it to you if you want to test it or not.

Notes:
 - I was not able to update the pipelines for an already created bootstrap Concourse for some reason, so if you need to do that:
    ```
    fly -t ${DEPLOY_ENV}-bootstrap get-pipeline -p create-bosh-concourse > pipeline.yml
    sed -i '' s/master/152065618_upgrade_concourse_4/ pipeline.yml
    fly -t ${DEPLOY_ENV}-bootstrap set-pipeline -p create-bosh-concourse -c pipeline.yml
    rm pipeline.yml
    ```

## Who can review

Not @bandesz
